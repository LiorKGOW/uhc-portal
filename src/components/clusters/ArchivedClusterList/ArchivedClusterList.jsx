/*
Copyright (c) 2019 Red Hat, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import size from 'lodash/size';
import isEmpty from 'lodash/isEmpty';
import PropTypes from 'prop-types';
import React, { Component } from 'react';

import { Spinner } from '@redhat-cloud-services/frontend-components';
import { LinkContainer } from 'react-router-bootstrap';
import { Link } from 'react-router-dom';

import {
  Alert,
  AlertActionCloseButton, Breadcrumb, BreadcrumbItem,
  Card,
  EmptyState,
  Split,
  SplitItem,
} from '@patternfly/react-core';

import ClusterListFilter from '../common/ClusterListFilter';
import ArchivedClusterListTable from './components/ArchiveClusterListTable/ArchivedClusterListTable';
import RefreshBtn from '../../common/RefreshButton/RefreshButton';
import ErrorTriangle from '../common/ErrorTriangle';
import GlobalErrorBox from '../common/GlobalErrorBox';


import ErrorBox from '../../common/ErrorBox';
import UnarchiveClusterDialog from '../common/UnarchiveClusterDialog';

import ViewPaginationRow from '../common/ViewPaginationRow/viewPaginationRow';
import helpers from '../../../common/helpers';
import { viewConstants } from '../../../redux/constants';


class ArchivedClusterList extends Component {
  constructor(props) {
    super(props);

    // refresh needs to be bound because it is passed to another component
    this.refresh = this.refresh.bind(this);
  }

  componentDidMount() {
    document.title = 'Archived Clusters | Red Hat OpenShift Cluster Manager';
    const {
      getCloudProviders, cloudProviders, setListFlag,
    } = this.props;

    setListFlag('showArchived', true);

    if (!cloudProviders.fulfilled && !cloudProviders.pending) {
      getCloudProviders();
    }
  }

  componentDidUpdate(prevProps) {
    // Check for changes resulting in a fetch
    const {
      viewOptions, valid, pending, unarchivedCluster, closeToast,
    } = this.props;
    if ((!valid && !pending)
      || helpers.viewPropsChanged(viewOptions, prevProps.viewOptions)) {
      this.refresh();
    }
    if (!prevProps.unarchivedCluster.showToast && unarchivedCluster.showToast) {
      setTimeout(closeToast, 8000);
    }
  }

  componentWillUnmount() {
    const { closeModal, invalidateClusters } = this.props;
    closeModal();
    invalidateClusters();
  }

  refresh() {
    const { fetchClusters, viewOptions } = this.props;
    fetchClusters(helpers.createViewQueryObject(viewOptions));
  }

  render() {
    const {
      error,
      valid,
      pending,
      clusters,
      viewOptions,
      setSorting,
      openModal,
      invalidateClusters,
      errorMessage,
      unarchivedCluster,
      operationID,
      closeToast,
    } = this.props;

    const toast = unarchivedCluster.showToast && (
      <Alert
        className="archived-cluster-toast"
        variant="success"
        title="Cluster successfully unarchived"
        action={<AlertActionCloseButton onClose={() => closeToast()} />}
      />
    );

    const breadCrumbs = (
      <Breadcrumb className="breadcrumbs-in-card">
        <LinkContainer to="">
          <BreadcrumbItem to="#">
            Clusters
          </BreadcrumbItem>
        </LinkContainer>
        <BreadcrumbItem isActive>
          Archived clusters
        </BreadcrumbItem>
      </Breadcrumb>
    );

    if (error && !size(clusters)) {
      return (
        <React.Fragment>
          {toast}
          <EmptyState>
            <ErrorBox
              message="Error retrieving archived clusters"
              response={{
                errorMessage,
                operationID,
              }}
            />
          </EmptyState>
        </React.Fragment>);
    }

    if ((!size(clusters) && pending && (isEmpty(viewOptions.filter) || !valid))) {
      return (
        <React.Fragment>
          {toast}
          <Card>
            <div className="cluster-list pf-c-content ocm-page">
              {breadCrumbs}
              <h1>Archived Clusters</h1>
              <div className="cluster-loading-container">
                <Spinner centered />
              </div>
            </div>
          </Card>
        </React.Fragment>
      );
    }

    return (
      <React.Fragment>
        {toast}
        <Card>
          <div className="cluster-list pf-c-content ocm-page">
            {breadCrumbs}
            <GlobalErrorBox />
            <h1>Archived Clusters</h1>
            <Split id="cluster-list-top">
              <SplitItem>
                <ClusterListFilter view={viewConstants.ARCHIVED_CLUSTERS_VIEW} />
              </SplitItem>
              <SplitItem className="pf-l-split__item split-margin-left">
                <div className="show-active-clusters-link">
                  <Link to="/">
                    Show active clusters
                  </Link>
                </div>
              </SplitItem>
              <SplitItem className="spinner-fit-container">
                { pending && <Spinner className="cluster-list-spinner" /> }
                { error && <ErrorTriangle errorMessage={errorMessage} className="cluster-list-warning" /> }
              </SplitItem>
              <SplitItem isFilled />
              <SplitItem>
                <ViewPaginationRow
                  viewType={viewConstants.ARCHIVED_CLUSTERS_VIEW}
                  currentPage={viewOptions.currentPage}
                  pageSize={viewOptions.pageSize}
                  totalCount={viewOptions.totalCount}
                  totalPages={viewOptions.totalPages}
                  variant="top"
                />
              </SplitItem>
              <SplitItem>
                <RefreshBtn autoRefresh refreshFunc={this.refresh} classOptions="cluster-list-top" />
              </SplitItem>
            </Split>
            <ArchivedClusterListTable
              clusters={clusters || []}
              viewOptions={viewOptions}
              setSorting={setSorting}
              openModal={openModal}
            />
            <ViewPaginationRow
              viewType={viewConstants.ARCHIVED_CLUSTERS_VIEW}
              currentPage={viewOptions.currentPage}
              pageSize={viewOptions.pageSize}
              totalCount={viewOptions.totalCount}
              totalPages={viewOptions.totalPages}
              variant="bottom"
            />
            <UnarchiveClusterDialog onClose={invalidateClusters} />
          </div>
        </Card>
      </React.Fragment>
    );
  }
}

ArchivedClusterList.propTypes = {
  invalidateClusters: PropTypes.func.isRequired,
  fetchClusters: PropTypes.func.isRequired,
  valid: PropTypes.bool.isRequired,
  clusters: PropTypes.array.isRequired,
  error: PropTypes.bool.isRequired,
  errorMessage: PropTypes.string.isRequired,
  pending: PropTypes.bool.isRequired,
  viewOptions: PropTypes.object.isRequired,
  setSorting: PropTypes.func.isRequired,
  getCloudProviders: PropTypes.func.isRequired,
  cloudProviders: PropTypes.object.isRequired,
  openModal: PropTypes.func.isRequired,
  closeModal: PropTypes.func.isRequired,
  setListFlag: PropTypes.func.isRequired,
  operationID: PropTypes.string,
  unarchivedCluster: PropTypes.shape({
    showToast: PropTypes.bool.isRequired,
  }),
  closeToast: PropTypes.func.isRequired,
};

export default ArchivedClusterList;
