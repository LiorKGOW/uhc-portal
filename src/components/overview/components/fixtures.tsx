import React from 'react';

import SomeLogo from '~/styles/images/RedHatOpenShiftGitOpsLogo.svg';

import { DRAWER_PANEL_CONTENT, DrawerPanelContentNode } from './common/DrawerPanelContent';
import PRODUCT_CARD_LOGOS from './common/ProductCardLogos';
import { FEATURED_PRODUCTS_CARDS } from './FeaturedProductsCards/FeaturedProductsCards';

const BASIC_PRODUCT_CARD_PROPS = {
  title: 'example title',
  description: 'example description',
  logo: 'product-overview-card__logo',
  labelText: 'example labelText',
  drawerPanelContent: {
    head: <div data-testid="test-head-div">example drawer panel content head</div>,
    body: <div data-testid="test-body-div">example drawer panel content body</div>,
  },
  openLearnMore: jest.fn(),
  isSelected: false,
  dataTestId: 'some-id',
};

const PRODUCT_CARD_TEST_CASES = {
  UNSELECTED: {
    ...BASIC_PRODUCT_CARD_PROPS,
  },

  SELECTED: {
    ...BASIC_PRODUCT_CARD_PROPS,
    isSelected: true,
  },

  WITHOUT_LABEL_TEXT: {
    ...BASIC_PRODUCT_CARD_PROPS,
    labelText: undefined,
  },
};

const BASIC_DRAWER_PANEL_PROPS = {
  children: <div data-testid="children-of-drawer-panel">some children</div>,
  content: {
    head: <div data-testid="test-head-div">example drawer panel content head</div>,
    body: <div data-testid="test-body-div">example drawer panel content body</div>,
  },
  isOpen: true,
  onClose: jest.fn(),
};

const DRAWER_PANEL_TEST_CASES = {
  OPENED: { ...BASIC_DRAWER_PANEL_PROPS },
};

const BASIC_RECCOMENDED_OPERATORS_CARDS_PROPS = {
  openLearnMore: jest.fn(),
  selectedCardTitle: '',
};

const RECOMMENDED_OPERATORS_CARDS_TEST_CASES = {
  NON_SELECTED: {
    ...BASIC_RECCOMENDED_OPERATORS_CARDS_PROPS,
  },
  GITOPS_SELECTED: {
    ...BASIC_RECCOMENDED_OPERATORS_CARDS_PROPS,
    selectedCardTitle: PRODUCT_CARD_LOGOS.gitops.title,
  },
};

type RecommendedOperatorsCardsNode = {
  title: string;
  description: string;
  logo: string | undefined;
  labelText?: string;
  drawerPanelContent?: DrawerPanelContentNode;
};

const RECOMMENDED_OPERATORS_CARDS_DATA: RecommendedOperatorsCardsNode[] = [
  {
    ...PRODUCT_CARD_LOGOS.gitops,
    description:
      'Integrate git repositories, continuous integration/continuous delivery (CI/CD) tools, and Kubernetes.',
    labelText: 'Free',
    drawerPanelContent: DRAWER_PANEL_CONTENT.gitops,
  },
  {
    ...PRODUCT_CARD_LOGOS.pipelines,
    description:
      'Automate your application delivery using a continuous integration and continuous deployment (CI/CD) framework.',
    labelText: 'Free',
    drawerPanelContent: DRAWER_PANEL_CONTENT.pipelines,
  },
  {
    ...PRODUCT_CARD_LOGOS.serviceMesh,
    description: 'Connect, manage, and observe microservices-based applications in a uniform way.',
    labelText: 'Free',
    drawerPanelContent: DRAWER_PANEL_CONTENT.serviceMesh,
  },
];

const BASIC_FEATURED_PRODUCTS_CARDS_PROPS = {
  openLearnMore: jest.fn(),
  selectedCardTitle: '',
};

const FEATURED_PRODUCTS_CARDS_TEST_CASES = {
  NON_SELECTED: {
    ...BASIC_FEATURED_PRODUCTS_CARDS_PROPS,
  },
  ADVANCED_CLUSTER_SECURITY_SELECTED: {
    ...BASIC_FEATURED_PRODUCTS_CARDS_PROPS,
    selectedCardTitle: PRODUCT_CARD_LOGOS.advancedClusterSecurity.title,
  },
};

const GENERAL_DRAWER_PANEL_HEAD_BASIC = {
  title: 'some title',
  logo: SomeLogo,
};

const GENERAL_DRAWER_PANEL_HEAD_WITH_LINK = {
  ...GENERAL_DRAWER_PANEL_HEAD_BASIC,
  trialButtonLink: 'https://www.redhat.com/en',
};

export {
  PRODUCT_CARD_TEST_CASES,
  DRAWER_PANEL_TEST_CASES,
  RECOMMENDED_OPERATORS_CARDS_TEST_CASES,
  RECOMMENDED_OPERATORS_CARDS_DATA,
  FEATURED_PRODUCTS_CARDS_TEST_CASES,
  FEATURED_PRODUCTS_CARDS,
  GENERAL_DRAWER_PANEL_HEAD_BASIC,
  GENERAL_DRAWER_PANEL_HEAD_WITH_LINK,
};
