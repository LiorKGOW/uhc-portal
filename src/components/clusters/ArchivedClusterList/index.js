import { connect } from 'react-redux';
import ArchivedClusterList from './ArchivedClusterList';

import { clustersActions } from '../../../redux/actions/clustersActions';
import { cloudProviderActions } from '../../../redux/actions/cloudProviderActions';
import { viewConstants } from '../../../redux/constants';
import { viewActions } from '../../../redux/actions/viewOptionsActions';
import { modalActions } from '../../common/Modal/ModalActions';
import { clearGlobalError } from '../../../redux/actions/globalErrorActions';

const mapDispatchToProps = {
  invalidateClusters: () => clustersActions.invalidateClusters(),
  fetchClusters: queryObj => clustersActions.fetchClusters(queryObj),
  setSorting: sorting => viewActions.onListSortBy(sorting, viewConstants.ARCHIVED_CLUSTERS_VIEW),
  setListFlag: (key, value) => viewActions.onListFlagsSet(key, value,
    viewConstants.ARCHIVED_CLUSTERS_VIEW),
  getCloudProviders: cloudProviderActions.getCloudProviders,
  openModal: modalActions.openModal,
  closeModal: modalActions.closeModal,
  clearGlobalError,
};

const mapStateToProps = state => ({
  ...state.clusters.clusters,
  username: state.userProfile.keycloakProfile.username,
  viewOptions: state.viewOptions[viewConstants.ARCHIVED_CLUSTERS_VIEW],
  cloudProviders: state.cloudProviders,
  sortByIndex: state.sortByIndex,
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(ArchivedClusterList);
