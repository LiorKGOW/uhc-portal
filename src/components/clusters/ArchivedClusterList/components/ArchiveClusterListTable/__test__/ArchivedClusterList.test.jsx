import React from 'react';
import { shallow } from 'enzyme';

import * as Fixtures from './ArchivedClusterList.fixtures';
import ArchivedClusterList from '../../../ArchivedClusterList';

describe('<ArchivedClusterList />', () => {
  describe('ArchivedClusterList', () => {
    const wrapper = shallow(
      <ArchivedClusterList
        viewOptions={{
          flags: {},
          fields: {},
        }}
        {...Fixtures}
      />,
    );

    it('should render', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
