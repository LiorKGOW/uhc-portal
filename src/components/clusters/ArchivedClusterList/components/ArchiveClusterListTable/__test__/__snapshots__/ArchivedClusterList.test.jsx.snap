// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ArchivedClusterList /> ArchivedClusterList should render 1`] = `
<Fragment>
  <PageHeader>
    <Breadcrumbs
      path={
        Array [
          Object {
            "label": "Clusters",
          },
          Object {
            "label": "Cluster Archives",
          },
        ]
      }
    />
    <PageHeaderTitle
      title="Cluster Archives"
    />
  </PageHeader>
  <PageSection>
    <Card>
      <div
        className="cluster-list"
      >
        <Connect(GlobalErrorBox) />
        <Split
          id="cluster-list-top"
        >
          <SplitItem>
            <Connect(ClusterListFilter)
              view="ARCHIVED_CLUSTERS_VIEW"
            />
          </SplitItem>
          <SplitItem
            className="pf-l-split__item split-margin-left"
          >
            <Connect(ClusterListFilterDropdown)
              isDisabled={false}
              view="ARCHIVED_CLUSTERS_VIEW"
            />
          </SplitItem>
          <SplitItem
            className="pf-l-split__item split-margin-left"
          >
            <div
              className="show-active-clusters-link"
            >
              <Link
                to="/"
              >
                Show active clusters
              </Link>
            </div>
          </SplitItem>
          <SplitItem
            className="spinner-fit-container"
          />
          <SplitItem
            isFilled={true}
          />
          <SplitItem>
            <Connect(ViewPaginationRow)
              variant="top"
              viewType="ARCHIVED_CLUSTERS_VIEW"
            />
          </SplitItem>
          <SplitItem>
            <RefreshBtn
              autoRefresh={true}
              classOptions="cluster-list-top"
              refreshFunc={[Function]}
            />
          </SplitItem>
        </Split>
        <Connect(ClusterListFilterChipGroup)
          view="ARCHIVED_CLUSTERS_VIEW"
        />
        <ArchivedClusterListTable
          clusters={
            Array [
              Object {
                "cluster_id": "1HAtdkNPWql68fuOI7KvuyM4OTp",
                "created_at": "2019-02-14T15:07:24.640289Z",
                "creator": Object {
                  "email": "***REMOVED***",
                  "href": "/api/accounts_mgmt/v1/accounts/1G1ikKt3haukUjrUDvRVLZyfEnY",
                  "id": "1G1ikKt3haukUjrUDvRVLZyfEnY",
                  "kind": "Account",
                  "name": "Elad Tabak",
                  "username": "etabak_privileged_uhc_20012019",
                },
                "href": "/api/accounts_mgmt/v1/subscriptions/1HAtguRKqqlQYCSFk14uwMl6g6p",
                "id": "1HAtguRKqqlQYCSFk14uwMl6g6p",
                "kind": "Subscription",
                "last_telemetry_date": "0001-01-01T00:00:00Z",
                "organization_id": "1HAIjRf0KAoWykFD2gexUhfngd0",
                "plan": Object {
                  "href": "/api/accounts_mgmt/v1/plans/OCP",
                  "id": "OCP",
                  "kind": "Plan",
                },
                "subscription": Object {
                  "id": "1HAtdkNPWql68fuOI7Kvuyk4OTp",
                  "last_reconcile_date": "0001-01-01T00:00:00Z",
                  "last_released_at": "0001-01-01T00:00:00Z",
                  "managed": false,
                  "provenance": "Provisioning",
                  "status": "Deprovisioned",
                },
                "support_level": "Eval",
                "updated_at": "2019-09-24T01:37:42.930173Z",
              },
            ]
          }
          openModal={[MockFunction]}
          setSorting={[MockFunction]}
          viewOptions={
            Object {
              "fields": Object {},
              "flags": Object {},
            }
          }
        />
        <Connect(ViewPaginationRow)
          variant="bottom"
          viewType="ARCHIVED_CLUSTERS_VIEW"
        />
        <Connect(ConnectedModal)
          ModalComponent={
            Object {
              "$$typeof": Symbol(react.memo),
              "WrappedComponent": [Function],
              "compare": null,
              "displayName": "Connect(UnarchiveClusterDialog)",
              "modalName": "unarchive-cluster",
              "type": [Function],
            }
          }
          onClose={
            [MockFunction] {
              "calls": Array [
                Array [],
              ],
              "results": Array [
                Object {
                  "type": "return",
                  "value": undefined,
                },
              ],
            }
          }
        />
      </div>
    </Card>
  </PageSection>
</Fragment>
`;
